sdk/core/core-rest-pipeline/src/policies/retryPolicy.ts

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { PipelineResponse, PipelineRequest, SendRequest } from "../interfaces";
+import { PipelinePolicy } from "../pipeline";
+import { delay } from "../util/helpers";
+import { createClientLogger } from "@azure/logger";
+import { RetryStrategy, RetryStrategyState } from "../retryStrategies/retryStrategy";
+import { RestError } from "../restError";
+import { AbortError } from "@azure/abort-controller";
+
+const retryPolicyLogger = createClientLogger("core-rest-pipeline retryPolicy");
+const DEFAULT_MAX_RETRIES = 3;
+
+/**
+ * The programmatic identifier of the retryPolicy.
+ */
+const retryPolicyName = "retryPolicy";
+
+/**
+ * retryPolicy is a generic policy to enable retrying requests when certain conditions are met
+ */
+export function retryPolicy(...strategies: RetryStrategy[]): PipelinePolicy {
+  return {
+    name: retryPolicyName,
+    async sendRequest(request: PipelineRequest, next: SendRequest): Promise<PipelineResponse> {
+      let response: PipelineResponse | undefined;
+      let responseError: RestError | undefined;
+      let retryCount = -1;
+
+      const strategyState: RetryStrategyState[] = Array(strategies.length);
+
+      // eslint-disable-next-line no-constant-condition
+      retryRequest: while (true) {
+        retryCount += 1;
+        response = undefined;
+        responseError = undefined;
+
+        try {
+          retryPolicyLogger.info(
+            `Retry ${retryCount}: Attempting to send request`,
+            request.requestId
+          );
+          response = await next(request);
+          retryPolicyLogger.info(
+            `Retry ${retryCount}: Received a response from request`,
+            request.requestId
+          );
+        } catch (e) {
+          retryPolicyLogger.info(
+            `Retry ${retryCount}: Received an error from request`,
+            request.requestId
+          );
+          responseError = e as RestError;
+          if (responseError.name !== "RestError") {
+            throw responseError;
+          }
+          response = responseError.response;
+        }
+
+        retryPolicyLogger.info(
+          `Retry ${retryCount}: Processing ${strategies.length} retry strategies.`
+        );
+
+        for (const [i, strategy] of strategies.entries()) {
+          const strategyLogger = strategy.logger || retryPolicyLogger;
+          strategyLogger.info(`Retry ${retryCount}: Processing retry strategy ${strategy.name}.`);
+          let state: RetryStrategyState = {
+            ...strategyState[i],
+            retryCount,
+            response,
+            responseError
+          };
+
+          if (state.retryCount >= (state.maxRetries || DEFAULT_MAX_RETRIES)) {
```

Comment:
Should this be `??` to avoid not using `state.maxRetries` of 0? 
```suggestion
          if (state.retryCount >= (state.maxRetries ?? DEFAULT_MAX_RETRIES)) {
```

Reply:
I like that!

Reply:
I don’t think so! Why would 0 max retries be useful?

Reply:
yes, if you wish to disable retries. Clearly they could do this by messing with the pipeline, but I can imagine someone wishing to configure this on a per-request basis eventually...

Reply:
ok ok I’ll take it!